In [2]:
import pandas as pd
from ast import literal_eval

In [3]:
movies_metadata = pd.read_csv("data/movies_metadata.csv")
movies_metadata.shape

C:\Users\Admin\AppData\Local\Temp\ipykernel_12704\1855133637.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies_metadata = pd.read_csv("data/movies_metadata.csv")


(45466, 24)

In [5]:
movies_metadata.head()
movies_metadata['genres'].head()

0    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1    [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2    [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3    [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                       [{'id': 35, 'name': 'Comedy'}]
Name: genres, dtype: object

In [6]:
print(movies_metadata['genres'][0])

[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]


In [7]:
movies_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [8]:
links_small = pd.read_csv("data/links_small.csv")
links_small.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9125 entries, 0 to 9124
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  9125 non-null   int64  
 1   imdbId   9125 non-null   int64  
 2   tmdbId   9112 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 214.0 KB


In [9]:
links_small.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [ ]:
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
links_small.shape


In [20]:
links_small.info()

<class 'pandas.core.series.Series'>
Index: 9112 entries, 0 to 9124
Series name: tmdbId
Non-Null Count  Dtype
--------------  -----
9112 non-null   int64
dtypes: int64(1)
memory usage: 142.4 KB


In [22]:
movies_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [24]:
# movies_metadata['id'].isin(links_small.astype('str'))
movies_metadata_small= movies_metadata[movies_metadata['id'].isin(links_small.astype('str'))]
movies_metadata_small.shape

(9099, 24)

In [25]:
# movies = movies_metadata_small['title']
movies = movies_metadata_small[['title','popularity','genres','release_date']]
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30  
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15  
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22  
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22  
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10

In [33]:
movies.head(1)

title popularity                                             genres  \
0  Toy Story  21.946943  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   

  release_date  
0   1995-10-30

In [38]:
# print(movies.iloc[0]['genres'])
# print(movies.loc[0,'title'])
print(movies.loc[0,['title','genres']])

title                                             Toy Story
genres    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
Name: 0, dtype: object


In [42]:
movies['genres'].loc[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

##### [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
##### -> ['Animation','Comedy','Family']
##### -> Animation Comedy Family

In [43]:
df = pd.DataFrame({
    "math":[80,90,100],
    "eng":[75,85,95]
})

In [45]:
df.apply(lambda x:x.mean())

math    90.0
eng     85.0
dtype: float64

- genres 에서 [] 데이터는 제외

    movies['genres'].fillna('[]') \
- 문자열을 리스트데이터로 변환

    .apply(literal_eval) \
- 장르 데이터를 추출하여 리스트로 변경

    .apply(lambda x: sorted(i['name'] for i in x) if isinstance(x, list) else [] ) \
- 리스트데이터를 하나의 문자열로 변환

    .apply(lambda x:" ".join(x) if len(x)>0 else None )

In [66]:
# movies['genres'].eq('[]').value_counts()

# movies['genres'].fillna('[]').apply(literal_eval).apply(lambda x: 참 if 조건 else 거짓 )
# movies['genres'].fillna('[]').apply(literal_eval).apply(lambda x: sorted(표현식 for i in x) if isinstance(x, list) else [] )
# test_list = movies['genres'].fillna('[]') \
#     .apply(literal_eval) \
#     .apply(lambda x: sorted(i['name'] for i in x) if isinstance(x, list) else [] )

movies['str_genres'] = movies['genres'] \
    .apply(literal_eval) \
    .apply(lambda x: sorted(i['name'] for i in x) if isinstance(x, list) else [] ) \
    .apply(lambda x:" ".join(x) if len(x)>0 else None )



C:\Users\Admin\AppData\Local\Temp\ipykernel_12704\3984659537.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['str_genres'] = movies['genres'] \


In [67]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9099 entries, 0 to 45265
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   title         9099 non-null   object
 1   popularity    9099 non-null   object
 2   genres        9099 non-null   object
 3   release_date  9099 non-null   object
 4   str_genres    9064 non-null   object
dtypes: object(5)
memory usage: 684.6+ KB


In [69]:
movies[['genres','str_genres']].head()

genres                str_genres
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   Animation Comedy Family
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  Adventure Family Fantasy
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...            Comedy Romance
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...      Comedy Drama Romance
4                     [{'id': 35, 'name': 'Comedy'}]                    Comedy

In [70]:
movies['release_date'] = pd.to_datetime(movies['release_date'])

C:\Users\Admin\AppData\Local\Temp\ipykernel_12704\1211608248.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['release_date'] = pd.to_datetime(movies['release_date'])


In [71]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9099 entries, 0 to 45265
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   title         9099 non-null   object        
 1   popularity    9099 non-null   object        
 2   genres        9099 non-null   object        
 3   release_date  9099 non-null   datetime64[ns]
 4   str_genres    9064 non-null   object        
dtypes: datetime64[ns](1), object(4)
memory usage: 684.6+ KB


In [ ]:
movies['year'] = movies['release_date'].dt.year
movies[['release_date','year']].head()

C:\Users\Admin\AppData\Local\Temp\ipykernel_12704\923264684.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['year'] = movies['release_date'].dt.year


release_date  year
0   1995-10-30  1995
1   1995-12-15  1995
2   1995-12-22  1995
3   1995-12-22  1995
4   1995-02-10  1995

In [73]:
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  year  
0   Animation Comedy Family  1995  
1  Adventure Family Fantasy  1995  
2            Comedy Romance  1995  
3      Comedy Drama Romance  1995  
4                    Comedy  1995

###  데이터 결측치 제거

In [ ]:
# movies.isnull().sum() #35
# movies.shape[0]

( movies.isnull().sum() / movies.shape[0] )


title           0.000000
popularity      0.000000
genres          0.000000
release_date    0.000000
str_genres      0.003847
year            0.000000
dtype: float64

In [84]:
movies = movies.dropna()
movies = movies.reset_index(drop=True)

In [85]:
movies.isnull().sum()

title           0
popularity      0
genres          0
release_date    0
str_genres      0
year            0
dtype: int64

### BoW기반 Contents Based Filtering



In [86]:
from sklearn.feature_extraction.text import CountVectorizer

In [87]:
bow_vector = CountVectorizer()

In [89]:
genre_mat = bow_vector.fit_transform(movies['str_genres'])

In [104]:
dict(sorted(bow_vector.vocabulary_.items()))

{'action': 0,
 'adventure': 1,
 'animation': 2,
 'comedy': 3,
 'crime': 4,
 'documentary': 5,
 'drama': 6,
 'family': 7,
 'fantasy': 8,
 'fiction': 9,
 'foreign': 10,
 'history': 11,
 'horror': 12,
 'movie': 13,
 'music': 14,
 'mystery': 15,
 'romance': 16,
 'science': 17,
 'thriller': 18,
 'tv': 19,
 'war': 20,
 'western': 21}

In [117]:
type(genre_mat)
genre_mat.toarray()[0]

array([0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### 코사인 유사도 측정

In [111]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity_of_genre = cosine_similarity(genre_mat,genre_mat) # data:genre_mat

In [113]:
similarity_of_genre.shape

(9064, 9064)

In [116]:
print(similarity_of_genre)

[[1.         0.33333333 0.40824829 ... 0.51639778 0.47140452 0.40824829]
 [0.33333333 1.         0.         ... 0.77459667 0.70710678 0.        ]
 [0.40824829 0.         1.         ... 0.         0.         0.5       ]
 ...
 [0.51639778 0.77459667 0.         ... 1.         0.73029674 0.        ]
 [0.47140452 0.70710678 0.         ... 0.73029674 1.         0.        ]
 [0.40824829 0.         0.5        ... 0.         0.         1.        ]]


In [118]:
import numpy as np


In [119]:
a = np.array([30,40,51,47,40])
index = np.argsort(a)
index

array([0, 1, 4, 3, 2])

In [120]:
movie0_similarity = similarity_of_genre[0]
movie0_similarity

array([1.        , 0.33333333, 0.40824829, ..., 0.51639778, 0.47140452,
       0.40824829], shape=(9064,))

In [121]:
movies[['title','str_genres']].iloc[0:2,:]

title                str_genres
0  Toy Story   Animation Comedy Family
1    Jumanji  Adventure Family Fantasy

### 유사도가 높은 영화의 index로 정렬

In [122]:
sorted_similarity_of_genre = similarity_of_genre.argsort()
sorted_similarity_of_genre

array([[  38, 7848, 9060, ..., 2718, 7064,    0],
       [ 926, 9054, 9063, ..., 8327, 5010,    1],
       [  58,   73, 9059, ...,    6,   49,   60],
       ...,
       [9060,   60,   61, ..., 9061, 3358, 7904],
       [9040, 9054, 9060, ..., 3020, 4340, 4238],
       [3400, 9042, 9043, ..., 6537, 3404, 3394]], shape=(9064, 9064))

In [123]:
sorted_similarity_of_genre = sorted_similarity_of_genre[:,::-1]
sorted_similarity_of_genre

array([[   0, 7064, 2718, ..., 9060, 7848,   38],
       [   1, 5010, 8327, ..., 9063, 9054,  926],
       [  60,   49,    6, ..., 9059,   73,   58],
       ...,
       [7904, 3358, 9061, ...,   61,   60, 9060],
       [4238, 4340, 3020, ..., 9060, 9054, 9040],
       [3394, 3404, 6537, ..., 9043, 9042, 3400]], shape=(9064, 9064))

In [127]:
movies[['title','str_genres']].iloc[[0,38],:]

title               str_genres
0                  Toy Story  Animation Comedy Family
38  Cry, the Beloved Country                    Drama